pkg/common-controller/snapshot_controller_base.go

```diff
@@ -195,131 +180,153 @@ func (ctrl *csiSnapshotCommonController) enqueueContentWork(obj interface{}) {
-				// so that syncSnapshot can delete the snapshot, should it still exist in the
-				// cluster after it's been removed from the informer cache
-				klog.V(5).Infof("updating snapshot %q; snapshotClass may have already been removed", key)
-				ctrl.updateSnapshot(newSnapshot)
-			}
-			return false
-		}
-		if err != nil && !errors.IsNotFound(err) {
-			klog.V(2).Infof("error getting snapshot %q from informer: %v", key, err)
-			return false
-		}
-		// The snapshot is not in informer cache, the event must have been "delete"
-		vsObj, found, err := ctrl.snapshotStore.GetByKey(key)
-		if err != nil {
-			klog.V(2).Infof("error getting snapshot %q from cache: %v", key, err)
-			return false
-		}
-		if !found {
-			// The controller has already processed the delete event and
-			// deleted the snapshot from its cache
-			klog.V(2).Infof("deletion of snapshot %q was already processed", key)
-			return false
-		}
-		snapshot, ok := vsObj.(*crdv1.VolumeSnapshot)
-		if !ok {
-			klog.Errorf("expected vs, got %+v", vsObj)
-			return false
-		}
-		newSnapshot, err := ctrl.checkAndUpdateSnapshotClass(snapshot)
-		if err == nil || errors.IsNotFound(err) {
-			// We should still handle deletion events even if the VolumeSnapshotClass
-			// is not found in the cluster
-			klog.V(5).Infof("deleting snapshot %q; snapshotClass may have already been removed", key)
-			ctrl.deleteSnapshot(newSnapshot)
-		}
+func (ctrl *csiSnapshotCommonController) processNextSnapshot() bool {
+	keyObj, quit := ctrl.snapshotQueue.Get()
+	if quit {
 		return false
 	}
+	defer ctrl.snapshotQueue.Done(keyObj)
+
+	if err := ctrl.syncSnapshotByKey(keyObj.(string)); err != nil {
+		// Rather than wait for a full resync, re-add the key to the
+		// queue to be processed.
+		ctrl.snapshotQueue.AddRateLimited(keyObj)
+		klog.V(4).Infof("Failed to sync snapshot %q, will retry again: %v", keyObj.(string), err)
+		return true
+	}
 
-	for {
-		if quit := workFunc(); quit {
-			klog.Infof("snapshot worker queue shutting down")
-			return
+	// Finally, if no error occurs we Forget this item so it does not
+	// get queued again until another change happens.
+	ctrl.snapshotQueue.Forget(keyObj)
+	return true
+}
+
+func (ctrl *csiSnapshotCommonController) syncSnapshotByKey(key string) error {
+	klog.V(5).Infof("syncSnapshotByKey[%s]", key)
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	klog.V(5).Infof("snapshotWorker: snapshot namespace [%s] name [%s]", namespace, name)
+	if err != nil {
+		klog.Errorf("error getting namespace & name of snapshot %q to get snapshot from informer: %v", key, err)
+		return nil
+	}
+	snapshot, err := ctrl.snapshotLister.VolumeSnapshots(namespace).Get(name)
+	if err == nil {
+		// The volume snapshot still exists in informer cache, the event must have
+		// been add/update/sync
+		newSnapshot, err := ctrl.checkAndUpdateSnapshotClass(snapshot)
+		if err == nil || (newSnapshot.ObjectMeta.DeletionTimestamp != nil && errors.IsNotFound(err)) {
```

Comment:
why deletion timestamp is checked here? A syncSnapshot is needed regardless whether the snapshot is being deleted or not in this case where volumesnapshot class is not found?

Reply:
my question here is what should be the behavior if volume snapshot class has been deleted from the system, however the snapshot has not been deleted?
currently it fails into this line:
https://github.com/kubernetes-csi/external-snapshotter/blob/9dce9a452d429fc0a727b8f1e92a0fa203189868/pkg/common-controller/snapshot_controller_base.go#L231
where:
1. the error message is ambiguous, it can be either fail to get snapshot or fail to get snapshotclass
2. even if the snapshotclass is missing, sync snapshot is still needed, no?

Reply:
If snapshot is not deleted but snapshot class is deleted, it should not go to line 231 "if err != nil && !errors.IsNotFound(err) { ".  It should be line 222.  In this case syncSnapshot will be called thru line 227 below.

I'm adding another message to clearly state that snapshot class is deleted.

Reply:
Done

Reply:
For deletion, we need to proceed to process the snapshot deletion so that there won't be any leaking resources.

For other cases, we return error so it is added to the queue.  Unless if the error is corrected (the snapshot class name is corrected or snapshot class is created again), the controller will retry forever.